openpype/hosts/openrv/api/pipeline.py

```diff
@@ -0,0 +1,245 @@
+# -*- coding: utf-8 -*-
+import os
+import json
+from collections import OrderedDict
+
+import pyblish
+import rv
+
+from openpype.host import HostBase, ILoadHost, IWorkfileHost, IPublishHost
+from openpype.hosts.openrv import OPENRV_ROOT_DIR
+from openpype.pipeline import (
+    register_loader_plugin_path,
+    register_inventory_action_path,
+    register_creator_plugin_path,
+    AVALON_CONTAINER_ID,
+)
+
+PLUGINS_DIR = os.path.join(OPENRV_ROOT_DIR, "plugins")
+PUBLISH_PATH = os.path.join(PLUGINS_DIR, "publish")
+LOAD_PATH = os.path.join(PLUGINS_DIR, "load")
+CREATE_PATH = os.path.join(PLUGINS_DIR, "create")
+INVENTORY_PATH = os.path.join(PLUGINS_DIR, "inventory")
+
+OPENPYPE_ATTR_PREFIX = "openpype."
+JSON_PREFIX = "JSON:::"
+
+
+class OpenRVHost(HostBase, IWorkfileHost, ILoadHost, IPublishHost):
+    name = "openrv"
+
+    def __init__(self):
+        super(OpenRVHost, self).__init__()
+        self._op_events = {}
+
+    def install(self):
+        pyblish.api.register_plugin_path(PUBLISH_PATH)
+        pyblish.api.register_host("openrv")
+
+        register_loader_plugin_path(LOAD_PATH)
+        register_creator_plugin_path(CREATE_PATH)
+        register_inventory_action_path(INVENTORY_PATH)
+
+    def open_workfile(self, filepath):
+        return rv.commands.addSources([filepath])
+
+    def save_workfile(self, filepath=None):
+        return rv.commands.saveSession(filepath)
+
+    def work_root(self, session):
+        work_dir = session.get("AVALON_WORKDIR")
+        scene_dir = session.get("AVALON_SCENEDIR")
+        if scene_dir:
+            return os.path.join(work_dir, scene_dir)
+        else:
+            return work_dir
+
+    def get_current_workfile(self):
+        filename = rv.commands.sessionFileName()
+        if filename == "Untitled":
+            return
+        else:
+            return filename
+
+    def workfile_has_unsaved_changes(self):
+        # RV has `State.unsavedChanges` attribute however that appears to
+        # always return false and is never set to be true. As such, for now
+        # we always return False.
+        return False
+
+    def get_workfile_extensions(self):
+        return [".rv"]
+
+    def get_containers(self):
+        for container in get_containers():
+            yield container
+
+    def update_context_data(self, data, changes):
+        imprint("root", data, prefix=OPENPYPE_ATTR_PREFIX)
+
+    def get_context_data(self):
+        return read("root", prefix=OPENPYPE_ATTR_PREFIX)
```

Comment:
I'm now storing global data on root since that node name sounded sensible, but I'm not sure if that's what its intent should be. Maybe someone more knowledgeable of (open)RV can comment on whether that is good practice.